packages/@core.keystone/cache.js

```diff
@@ -0,0 +1,141 @@
+ *                           -> [GQL Request] -> [DB]
+ *
+ * On every incoming request module simply checks x-request-id header and operation.
+ * If request is `read` - return cache item (assuming cache has this item)
+ * If request has other type - remove cache item by request id
+ *
+ * Cache structure:
+ * cache: {
+ *     <x-request-id>: {
+ *      	<request-args>: []
+ *      	<request-args>: []
+ *      	<request-args>: []
+ *     }
+ * }
+ *
+ */
+
+const express = require('express')
+const { get } = require('lodash')
+
+class KeystoneCacheMiddleware {
+
+    constructor (cache) {
+        if (!(typeof cache === 'object') || cache === null) {
+            throw new Error('Cache should be dict!')
+        }
+        this.cache = cache
+    }
+
+    prepareMiddleware ({ keystone, dev, distDir }) {
+
+        const app = express()
+        app.use((req, res, next) => {
+            const requestId = get(req, ['headers', 'x-request-id'])
+            res.on('close', () => {
+                if (requestId) { delete this.cache[requestId] }
+            })
+            next()
+        })
+        return app
+    }
+}
+
+const generateRequestKey = (gqlName, args) => (
+    `${gqlName}-${JSON.stringify(args)}`
+)
+
+const getRequestIdFromContext = (context) => {
+    return get(context, ['req', 'headers', 'x-request-id'], null)
+}
+
+const patchMutation = (mutationContext, mutation, isUpdateMutation, cache) => {
+    return async (data, context, mutationState) => {
+        let requestId
+        if (isUpdateMutation) {
+            requestId = getRequestIdFromContext(mutationState)
+        } else {
+            requestId = getRequestIdFromContext(context)
+        }
+        if (requestId) { delete cache[requestId] }
+        return await mutation.call(mutationContext, data, context,
+            mutationState)
+    }
+}
+
+const patchQuery = (queryContext, query, cache) => {
+    return async function (args, context, gqlName, info, from) {
+        let key = null
+        const requestId = getRequestIdFromContext(context)
+
+        if (requestId) {
+            key = generateRequestKey(gqlName, args)
+
+            if (!(requestId in cache)) {
+                cache[requestId] = {}
+            }
+
+            // Drop the key, if the operation type is mutation
+            const operationType = get(info, ['operation', 'operation'])
+            if (operationType !== 'query') {
+                delete cache[requestId][key]
+            }
+
+            if (key in cache[requestId]) {
+                return cache[requestId][key]
+            }
+        }
+        const listResult = await query.call(queryContext, args, context, gqlName, info, from)
+
+        if (requestId && key) { cache[requestId][key] = listResult }
+
+        return listResult
+    }
+}
+
+const initCache = (keystone, cache) => {
+
+    const originalCreateList = keystone.createList
+
+    console.debug('INITIATING KEYSTONE-LEVEL-CACHE')
```

Comment:
?

Reply:
To actually see if the cache is enabled

Reply:
You can check it by state. It’s redundant log

Reply:
Done